base/src/main/java/proguard/util/CircularBuffer.java

```diff
@@ -0,0 +1,21 @@
+package proguard.util;
+
+import java.util.Stack;
+
+public class CircularBuffer<T> extends Stack<T> {
```

Comment:
Implementing a circular buffer as an extension of a stack is pretty inefficient -- adding an element has O(N) complexity.

Reply:
Oh yes! Seems like it, good catch. Let's rewrite that with an int[] ourselves :D